### PR TITLE
libs\platform\user\nmr_impl.cpp should implement complete NMR state machine

### LINKED ISSUE #2048: libs\platform\user\nmr_impl.cpp should implement complete NMR state machine
The code in libs\platform\user\nmr_impl.cpp currently misses several possible state transitions that client/provider could make.

Examples are:
1) If the client fails after calling NmrClientAttachProvider and having it return success.
2) Should call ProviderDetachClient if ProviderAttachClient succeeds, but client rejects the attach.

Currently this is not an issue as the existing test code doesn't exercise these paths, but for completeness they should be supported.

### PR DESCRIPTION
**Pull request by delaramamiri**
Closes #2048 
## Description

The code in libs\platform\user\nmr_impl.cpp currently misses several possible state transitions that client/provider could make.

Examples are:

If the client fails after calling NmrClientAttachProvider and having it return success.
Should call ProviderDetachClient if ProviderAttachClient succeeds, but client rejects the attach.

### CODE REVIEW

**Review by dthaler — CHANGES_REQUESTED** [review]
I notice the PR description is missing the "Tests" section, and the changed files do not include any test changes.  Please add one or more test cases to exercise this PR.

File: libs/platform/unit/platform_unit_test.cpp [hunk 1]
```diff
@@ -472,6 +536,43 @@ TEST_CASE("extension_test", "[platform]")
...
+            &interface_id,
+            &provider_module_id,
+            nullptr,
+            &provider_data,
+            &test_provider_dispatch_table,
+            &callback_context,
+            (NPI_PROVIDER_ATTACH_CLIENT_FN*)test_failed_provider_attach_client,
+            (NPI_PROVIDER_DETACH_CLIENT_FN*)test_provider_detach_client,
+            nullptr) == EBPF_SUCCESS);
+    REQUIRE(
```
Comment (Alan-Jowett) [inline libs/platform/unit/platform_unit_test.cpp:564]: How are you validating that you get the correct behavior from NMR here?

File: libs/platform/unit/platform_unit_test.cpp [hunk 2]
```diff
@@ -440,6 +441,73 @@ test_provider_attach_client(
     return STATUS_SUCCESS;
 };
 
+static NTSTATUS
+test_denied_client_attach_provider(
```
Comment (shankarseal) [inline libs/platform/unit/platform_unit_test.cpp:445]: consider renaming to `test_denied_provider_attach_client`

File: libs/platform/unit/platform_unit_test.cpp (hunk 1 above)
Comment (shankarseal) [reply libs/platform/unit/platform_unit_test.cpp:575]: I think Alan means, that for each of the provider_load (NMR provder) you will also have to call extension_load (NMR client) - and check that it is returning the same error code that the test provider is returning.

File: libs/platform/unit/platform_unit_test.cpp [hunk 3]
```diff
@@ -440,6 +441,73 @@ test_provider_attach_client(
...
+    UNREFERENCED_PARAMETER(client_binding_context);
+    UNREFERENCED_PARAMETER(client_dispatch);
+    *provider_binding_context = nullptr;
+    *provider_dispatch = &test_provider_dispatch_table;
+    test_denied_attach_called = true;
+    return STATUS_ACCESS_DENIED;
+};
+
+static NTSTATUS
+test_failed_provider_register_client(
```
Comment (shankarseal) [inline libs/platform/unit/platform_unit_test.cpp:489]: This seems to be identical to the previous function. Did you mean to return a different failure code?

File: libs/platform/unit/platform_unit_test.cpp (hunk 2 above)
Comment (delaramamiri) [reply libs/platform/unit/platform_unit_test.cpp]: Modified.

**Pull request closed without merging**